src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,140 @@
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                co_await item.client->stop();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        // Grab the node located at that position
+        auto node = inner_hash.extract(it_hash);
+
+        // Reuse the client if the passwords match
+        // Otherwise make a new client
+        if (node.value().client->config().scram_password.value() == user.pass) {
+            vlog(plog.debug, "Reuse client for user {}", k);
+        } else {
+            vlog(plog.debug, "Password changed. Remake client for user {}", k);
+            client_ptr old = node.value().client;
+            bool stop_old_client = false;
+            client_ptr next;
+            try {
+                next = make_client(user, authn_method);
+                // Try connect as a safety check
+                co_await next->connect();
+            } catch (const std::exception& ex) {
+                node.value().client = next;
+                stop_old_client = true;
+            }
+
+            if (stop_old_client) {
+                co_await old->stop();
+            }
+        }
+
+        // Update the timestamp and put it to the front
+        node.value().last_used = timestamped_user::clock::now();
+        client = node.value().client;
+        inner_list.insert(inner_list.begin(), std::move(node));
+    }
+
+    co_return client;
+}
+
+// Modified boost remove_if so we can call kafka::client::stop
+// See sequenced_index_remove in
+// https://github.com/boostorg/multi_index/blob/develop/include/boost/multi_index/detail/seq_index_ops.hpp
+ss::future<> kafka_client_cache::clean_stale_clients() {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto has_expired = timestamped_user::clock::now();
+    auto condition = [this, has_expired](const timestamped_user& item) {
+        return has_expired >= (item.last_used + _keep_alive());
+    };
+
+    auto first = inner_list.begin();
+    auto last = inner_list.end();
+    while (first != last) {
+        if (condition(*first)) {
+            vlog(plog.debug, "Remove {} from cache", first->key);
+            // Stop client before erase
+            co_await first->client->stop();
```

Comment:
looks like this races with stopping the client in fetch_or_insert

Reply:
I'll re-work so GC is the only place where clients are stopped.

Reply:
you can also stop them here. it makes sense if you are evicting them on demand. it just needs to be safe.

Reply:
https://github.com/redpanda-data/redpanda/pull/6452#discussion_r982929252